### ISSUE TITLE
llm-server crashes after "Measuring VRAM headroom"

### ISSUE THREAD

**Issue by janvitos** [raketenkater/ggrun#7]
GPU: RTX 4070 12 GB
RAM: 32 GB
OS: WSL2
llama version: ik_llama.cpp (latest)

Hello,

Thanks for this great promising tool! I really look forward to getting it working on my PC.

I am facing an issue: When running llm-server google_gemma-4-26B-A4B-it-IQ3_XXS.gguf --ai-tune, the tool crahes right after "Measuring VRAM headroom..." Unfortunately, there's no error message, so it's hard to debug.

I am going to test with llama.cpp backend instead of ik_llama.cpp and see if it helps.

Edit: Tried with llama as a backend. Around Round 3/8, the tool crashed again and froze my entire PC. Had to hard reset. First time my PC freezes every since I built it 2 years ago.

**Comment by raketenkater** [comment]
ok i put a few oom protections in place especially for wsl so your system wont freeze (i hate those freezes) 

 could you test again if it now works ?

**Comment by janvitos** [comment]
Certainly. I will be testing with llama.cpp, just so you know.

**Comment by janvitos** [comment]
Didn't even make it out of round 0/8 unfortunately (error below).

And just so you know, I can load this model fine and use it with opencode when using standalone llama-server.

To me, it seems like the http server was not properly shutdown after the first "pre-pass" and was still running when it tried to do the first pass. So it couldn't bind port 8081. I had nothing running on that port, so it must be a llm-server self conflict.

═══════════════════════════════════════════
  Model: google_gemma-4-26B-A4B-it-IQ3_XXS.gguf
  Size: 11.3GB | Layers: 30 | Experts: 128
  GPU layers: 18 | CPU layers: 12
  Context: 65536 | Port: 8081
  Config cached: /xxxx/xxxx/.cache/llm-server/45753e6ea5c5b369a711ba207d262ebb.conf
═══════════════════════════════════════════

MoE placement complete. Running AI Tune with placement locked...

═══════════════════════════════════════════════════
  AI Tune — iterative LLM-driven optimization
  Model: google_gemma-4-26B-A4B-it-IQ3_XXS.gguf
  Rounds: 8
═══════════════════════════════════════════════════

Round 0/8: Benchmarking baseline (heuristic config)...
ggml_cuda_init: found 1 CUDA devices (Total VRAM: 12281 MiB):
  Device 0: NVIDIA GeForce RTX 4070 SUPER, compute capability 8.9, VMM: yes, VRAM: 12281 MiB
main: n_parallel is set to auto, using n_parallel = 4 and kv_unified = true
build_info: b8807-b3d758750
system_info: n_threads = 8 (n_threads_batch = 8) / 16 | CUDA : ARCHS = 890 | USE_GRAPHS = 1 | PEER_MAX_BATCH_SIZE = 128 | CPU : SSE3 = 1 | SSSE3 = 1 | AVX = 1 | AVX_VNNI = 1 | AVX2 = 1 | F16C = 1 | FMA = 1 | BMI2 = 1 | AVX512 = 1 | AVX512_VBMI = 1 | AVX512_VNNI = 1 | AVX512_BF16 = 1 | LLAMAFILE = 1 | OPENMP = 1 | REPACK = 1 |
init: using 15 threads for HTTP server
start: binding port with default address family
start: couldn't bind HTTP server socket, hostname: 0.0.0.0, port: 8081
srv    operator(): operator(): cleaning up before exit...
main: exiting due to HTTP server error
ERROR: Baseline config failed to start

EDIT: Currently at Round 2/8, so it seems the http server bug is intermittent. Might be worth checking into that.

EDIT: Got another crash (ends at Reason:)

Round 2/8: Querying model for config proposal...
srv  params_from_: Chat format: peg-gemma4
slot get_availabl: id  3 | task -1 | selected slot by LRU, t_last = -1
srv  get_availabl: updating prompt cache
srv          load:  - looking for better prompt, base f_keep = -1.000, sim = 0.000
srv        update:  - cache state: 0 prompts, 0.000 MiB (limits: 8192.000 MiB, 65536 tokens, 8589934592 est)
srv  get_availabl: prompt cache update took 0.01 ms
slot launch_slot_: id  3 | task -1 | sampler chain: logits -> ?penalties -> ?dry -> ?top-n-sigma -> top-k -> ?typical -> top-p -> min-p -> ?xtc -> temp-ext -> dist
slot launch_slot_: id  3 | task 0 | processing task, is_child = 0
slot update_slots: id  3 | task 0 | new prompt, n_ctx_slot = 65536, n_keep = 0, task.n_tokens = 12794
slot update_slots: id  3 | task 0 | n_tokens = 0, memory_seq_rm [0, end)
slot update_slots: id  3 | task 0 | prompt processing progress, n_tokens = 2048, batch.n_tokens = 2048, progress = 0.160075
slot update_slots: id  3 | task 0 | n_tokens = 2048, memory_seq_rm [2048, end)
slot update_slots: id  3 | task 0 | prompt processing progress, n_tokens = 4096, batch.n_tokens = 2048, progress = 0.320150
slot update_slots: id  3 | task 0 | n_tokens = 4096, memory_seq_rm [4096, end)
slot update_slots: id  3 | task 0 | prompt processing progress, n_tokens = 6144, batch.n_tokens = 2048, progress = 0.480225
slot update_slots: id  3 | task 0 | n_tokens = 6144, memory_seq_rm [6144, end)
slot update_slots: id  3 | task 0 | prompt processing progress, n_tokens = 8192, batch.n_tokens = 2048, progress = 0.640300
slot update_slots: id  3 | task 0 | n_tokens = 8192, memory_seq_rm [8192, end)
slot update_slots: id  3 | task 0 | 8192 tokens since last checkpoint at 0, creating new checkpoint during processing at position 10240
slot update_slots: id  3 | task 0 | prompt processing progress, n_tokens = 10240, batch.n_tokens = 2048, progress = 0.800375
slot update_slots: id  3 | task 0 | created context checkpoint 1 of 32 (pos_min = 3584, pos_max = 8191, n_tokens = 8192, size = 253.178 MiB)
slot update_slots: id  3 | task 0 | n_tokens = 10240, memory_seq_rm [10240, end)
slot update_slots: id  3 | task 0 | prompt processing progress, n_tokens = 12278, batch.n_tokens = 2038, progress = 0.959669
slot update_slots: id  3 | task 0 | n_tokens = 12278, memory_seq_rm [12278, end)
slot update_slots: id  3 | task 0 | prompt processing progress, n_tokens = 12790, batch.n_tokens = 512, progress = 0.999687
slot update_slots: id  3 | task 0 | created context checkpoint 2 of 32 (pos_min = 7670, pos_max = 12277, n_tokens = 12278, size = 253.178 MiB)
slot update_slots: id  3 | task 0 | n_tokens = 12790, memory_seq_rm [12790, end)
slot init_sampler: id  3 | task 0 | init sampler, took 1.04 ms, tokens: text = 12794, total = 12794
slot update_slots: id  3 | task 0 | prompt processing done, n_tokens = 12794, batch.n_tokens = 4
slot update_slots: id  3 | task 0 | created context checkpoint 3 of 32 (pos_min = 8182, pos_max = 12789, n_tokens = 12790, size = 253.178 MiB)
slot print_timing: id  3 | task 0 |
prompt eval time =    8364.54 ms / 12794 tokens (    0.65 ms per token,  1529.55 tokens per second)
       eval time =  516187.12 ms / 16384 tokens (   31.51 ms per token,    31.74 tokens per second)
      total time =  524551.66 ms / 29178 tokens
slot      release: id  3 | task 0 | stop processing: n_tokens = 29177, truncated = 0
srv  update_slots: all slots are idle
srv  log_server_r: done request: POST /v1/chat/completions 127.0.0.1 200
  Config: config_2
  Reason:

**Comment by janvitos** [comment]
I am going to test with CUDA 12.8 instead of 13.1 and report back.

EDIT: Same issue. Crahed with no errors after Round 2/8.

These are the last lines right before the crash:

prompt eval time =    7811.81 ms / 12524 tokens (    0.62 ms per token,  1603.21 tokens per second)
       eval time =  368227.82 ms / 16384 tokens (   22.47 ms per token,    44.49 tokens per second)
      total time =  376039.62 ms / 28908 tokens
slot      release: id  2 | task 6 | stop processing: n_tokens = 28907, truncated = 0
srv  update_slots: all slots are idle
srv  log_server_r: done request: POST /v1/chat/completions 127.0.0.1 200
  Config: config_2
  Reason:

**Comment by janvitos** [comment]
Was able to get --ai-tune working using the --tune-long-timeout flag.
